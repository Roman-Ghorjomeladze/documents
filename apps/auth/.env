JWT_SECRET=fdlsalfsdjflasjfaslfasojgkh;fsdfaurewqeu2342342
JWT_EXPIRES_IN=2d