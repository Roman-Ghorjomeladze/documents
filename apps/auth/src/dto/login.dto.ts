import { ApiProperty } from '@nestjs/swagger';
import { IsEmail, IsString, MaxLength } from 'class-validator';

export class LoginDto {
  @ApiProperty({ required: true })
  @MaxLength(60)
  @IsEmail()
  email: string;

  @ApiProperty({ required: true })
  @IsString()
  password: string;
}
