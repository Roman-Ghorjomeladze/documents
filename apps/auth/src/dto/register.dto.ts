import {
  IsEmail,
  IsNotEmpty,
  IsString,
  IsStrongPassword,
  MaxLength,
} from 'class-validator';
import { ApiProperty } from '@nestjs/swagger';

export class RegisterDto {
  @ApiProperty({ required: true })
  @IsString()
  @MaxLength(60)
  firstName: string;

  @ApiProperty({ required: true })
  @IsString()
  @MaxLength(60)
  lastName: string;

  @ApiProperty({ required: true })
  @MaxLength(60)
  @IsNotEmpty()
  @IsEmail()
  email: string;

  @ApiProperty({ required: true })
  @IsStrongPassword({
    minLength: 6,
    minUppercase: 1,
    minSymbols: 1,
    minNumbers: 1,
  })
  password: string;
}
