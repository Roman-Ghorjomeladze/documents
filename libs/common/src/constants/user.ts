export enum UserStatuses {
  ACTIVE = 'ACTIVE',
  PENDING = 'PENDING',
  REJECTED = 'REJECTED',
  DEACTIVATED = 'DEACTIVATED',
}
