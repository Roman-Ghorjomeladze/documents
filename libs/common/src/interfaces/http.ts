import { Request } from 'express';

export interface AuthorizedRequest extends Request {
  user: RequestUser;
}

export interface RequestUser {
  id: number;
  email: string;
}
