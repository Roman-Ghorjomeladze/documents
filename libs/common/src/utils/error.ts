export class ApiError extends Error {
  msg: string;
  status: number;
  type: string;
  constructor(msg: string, status: number, type: string) {
    super(msg);

    this.msg = msg;
    this.status = status;
    this.type = type;
  }

  toJSON() {
    return {
      message: this.message,
      trace: this.stack,
      status: this.status,
      type: this.type,
    };
  }
}
