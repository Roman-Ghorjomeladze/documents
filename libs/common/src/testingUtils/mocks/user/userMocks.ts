import { User } from '../../../../modules/users/entities/user.entity';
import { UserStatuses } from '../../../constants/user';

export const userMock: User = (() => {
  const user: User = new User();
  user.id = 1;
  user.firstName = 'John';
  user.lastName = 'Doe';
  user.email = 'john@doe.com';
  user.password = '123456';
  user.createdAt = new Date('2024-11-02T00:34:48.592Z');
  user.updatedAt = new Date('2024-11-02T00:34:48.592Z');
  user.deletedAt = null;
  user.status = UserStatuses.ACTIVE;
  return user;
})();
