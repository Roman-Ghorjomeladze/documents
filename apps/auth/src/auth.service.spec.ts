import { Test, TestingModule } from '@nestjs/testing';
import { AuthService } from './auth.service';
import { UsersService } from '../users/users.service';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { EmailIsTakenError } from '../users/errors/EmailIsTaken.error';
import { UserNotFoundByCredentialsError } from './errors/userNotFound.error';
import { RegisterDto } from './dto/register.dto';
import { LoginDto } from './dto/login.dto';
import { RequestUser } from '../../interfaces/http';
import { userMock } from '../../../common/testingUtils/mocks/user/userMocks';

describe('AuthService', () => {
  let authService: AuthService;
  let userService: UsersService;
  let jwtService: JwtService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AuthService,
        {
          provide: UsersService,
          useValue: {
            findByEmail: jest.fn(),
            createUser: jest.fn(),
            findOne: jest.fn(),
          },
        },
        {
          provide: JwtService,
          useValue: {
            sign: jest.fn(),
          },
        },
      ],
    }).compile();

    authService = module.get<AuthService>(AuthService);
    userService = module.get<UsersService>(UsersService);
    jwtService = module.get<JwtService>(JwtService);
  });

  describe('register', () => {
    it('should throw EmailIsTakenError if email is already taken', async () => {
      const registerDto: RegisterDto = {
        email: 'test@example.com',
        password: 'password123',
        firstName: 'John',
        lastName: 'Doe',
      };
      jest.spyOn(userService, 'findByEmail').mockResolvedValue(userMock);

      await expect(authService.register(registerDto)).rejects.toThrow(
        EmailIsTakenError,
      );
      expect(userService.findByEmail).toHaveBeenCalledWith(registerDto.email);
    });

    it('should create a new user if email is not taken', async () => {
      const registerDto: RegisterDto = {
        email: 'test@example.com',
        password: 'password123',
        firstName: 'John',
        lastName: 'Doe',
      };

      jest.spyOn(userService, 'findByEmail').mockResolvedValue(null);
      jest.spyOn(userService, 'createUser').mockResolvedValue(userMock);

      const result = await authService.register(registerDto);
      expect(userService.findByEmail).toHaveBeenCalledWith(registerDto.email);
      expect(userService.createUser).toHaveBeenCalledWith(registerDto);
      expect(result).toMatchSnapshot();
    });
  });

  describe('login', () => {
    it('should throw UserNotFoundByCredentialsError if user is not found', async () => {
      const loginDto: LoginDto = {
        email: 'test@example.com',
        password: 'password123',
      };
      jest.spyOn(userService, 'findByEmail').mockResolvedValue(null);

      await expect(authService.login(loginDto)).rejects.toThrow(
        UserNotFoundByCredentialsError,
      );
      expect(userService.findByEmail).toHaveBeenCalledWith(
        loginDto.email,
        true,
      );
    });

    it('should throw UserNotFoundByCredentialsError if passwords do not match', async () => {
      const loginDto: LoginDto = {
        email: 'test@example.com',
        password: 'password123',
      };

      jest.spyOn(userService, 'findByEmail').mockResolvedValue(userMock);
      jest.spyOn(bcrypt, 'compare').mockImplementation(async () => false);

      await expect(authService.login(loginDto)).rejects.toThrow(
        UserNotFoundByCredentialsError,
      );
      expect(bcrypt.compare).toHaveBeenCalledWith(
        loginDto.password,
        userMock.password,
      );
    });

    it('should return user and token if login is successful', async () => {
      const loginDto: LoginDto = {
        email: userMock.email,
        password: userMock.password,
      };
      const mockToken = 'jwt-token';

      jest.spyOn(userService, 'findByEmail').mockResolvedValue(userMock);
      jest.spyOn(bcrypt, 'compare').mockImplementation(async () => true);
      jest.spyOn(jwtService, 'sign').mockReturnValue(mockToken);

      const result = await authService.login(loginDto);
      expect(userService.findByEmail).toHaveBeenCalledWith(
        loginDto.email,
        true,
      );
      expect(bcrypt.compare).toHaveBeenCalledWith(
        loginDto.password,
        userMock.password,
      );
      expect(jwtService.sign).toHaveBeenCalledWith({
        id: userMock.id,
        email: userMock.email,
      });
      expect(result).toMatchSnapshot();
    });
  });

  describe('me', () => {
    it('should return user information', async () => {
      const mockUser: RequestUser = { id: userMock.id, email: userMock.email };

      jest.spyOn(userService, 'findOne').mockResolvedValue(userMock);

      const result = await authService.me(mockUser);
      expect(userService.findOne).toHaveBeenCalledWith(mockUser.id);
      expect(result).toMatchSnapshot();
    });
  });
});
