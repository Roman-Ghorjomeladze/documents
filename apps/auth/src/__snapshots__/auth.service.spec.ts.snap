// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthService login should return user and token if login is successful 1`] = `
{
  "token": "jwt-token",
  "user": User {
    "createdAt": 2024-11-02T00:34:48.592Z,
    "deletedAt": null,
    "email": "john@doe.com",
    "firstName": "John",
    "id": 1,
    "lastName": "Doe",
    "password": "123456",
    "status": "ACTIVE",
    "updatedAt": 2024-11-02T00:34:48.592Z,
  },
}
`;

exports[`AuthService me should return user information 1`] = `
User {
  "createdAt": 2024-11-02T00:34:48.592Z,
  "deletedAt": null,
  "email": "john@doe.com",
  "firstName": "John",
  "id": 1,
  "lastName": "Doe",
  "password": "123456",
  "status": "ACTIVE",
  "updatedAt": 2024-11-02T00:34:48.592Z,
}
`;

exports[`AuthService register should create a new user if email is not taken 1`] = `
User {
  "createdAt": 2024-11-02T00:34:48.592Z,
  "deletedAt": null,
  "email": "john@doe.com",
  "firstName": "John",
  "id": 1,
  "lastName": "Doe",
  "password": "123456",
  "status": "ACTIVE",
  "updatedAt": 2024-11-02T00:34:48.592Z,
}
`;
